windowsforms/FlexGridShowcaseDemo/cs/MainForm.cs

```diff
@@ -0,0 +1,682 @@
+        private DataSet _dataSet = new DataSet();
+        private C1RulesManager _rulesManager = new C1RulesManager();
+        private IEnumerable<IRule> _rules = Enumerable.Empty<IRule>();
+        private static Random _rnd = new Random();
+
+        #region Initialization
+
+        public MainForm()
+        {
+            InitializeComponent();
+        }
+
+        private void MainForm_Load(object sender, EventArgs e)
+        {
+            FillData();
+
+            InitImages();
+            InitFlexGrid();
+            InitRules();
+            InitThemes();
+
+            // Data sizes
+            var dataSizes = new List<string>() { "10 Rows, 12 Columns", "50 Rows, 12 Columns", "100 Rows, 12 Columns", "1000 Rows, 12 Columns", "5000 Rows, 12 Columns" };
+            foreach (var item in dataSizes)
+            {
+                _ribbonComboBoxDataSize.Items.Add(item);
+            }
+
+            // Columns visible
+            _ribbonMenuColumns.Items.Clear();
+            var columnRibbonToggleButtons = (from s in _flexGrid.Cols.Cast<Column>() select s)
+                .Where(x => !string.IsNullOrEmpty(x.Name))
+                .Select(x => new RibbonToggleButton()
+                {
+                    Text = x.Caption,
+                    Pressed = x.Visible
+                });
+
+            foreach (var columnRibbonToggleButton in columnRibbonToggleButtons)
+            {
+                columnRibbonToggleButton.PressedButtonChanged += ColumnRibbonToggleButton_PressedButtonChanged;
+                _ribbonMenuColumns.Items.Add(columnRibbonToggleButton);
+            }
+
+            _ribbonComboBoxDataSize.SelectedIndex = 2;
+
+            ActiveControl = _flexGrid;
+        }
+
+        #endregion
+
+        #region Fill Data
+
+        private enum Countries
+        {
+            Germany,
+            UK,
+            US,
+            Japan
+        };
+
+        private enum DrawColors
+        {
+            Black,
+            White,
+            Green,
+            Red
+        };
+
+        private void BuildRows(int rowCount, DataTable dataTable)
+        {
+            var startPeriod = new DateTime(2020, 01, 25, 8, 29, 0);
+            _flexGrid.BeginUpdate();
+
+            // Related data
+            var products = (from s in _dataSet.Tables["Products"].Rows.Cast<DataRow>() select s)
+                    .Select(x => x["Name"].ToString())
+                    .ToArray();
+
+            var countries = Enum.GetValues(typeof(Countries))
+                    .Cast<Countries>()
+                    .ToArray();
+
+            var colors = Enum.GetValues(typeof(DrawColors))
+                    .Cast<DrawColors>()
+                    .ToArray();
+
+            dataTable.Clear();
+
+            var dataRows = dataTable.Rows;
+            for (var rowIndex = 0; rowIndex < rowCount; rowIndex++)
+            {
+                // Fill history data
+                var historyData = new int[12];
```

Comment:
Not sure if stackalloc can be used here or not :-(

Reply:
Don't stackalloc in loops: https://docs.microsoft.com/visualstudio/code-quality/ca2014